sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -381,6 +383,16 @@ class Dataset[T] private[sql](
     new Dataset(sqlContext, execution, ExpressionEncoder.tuple(encoders))
   }
 
+  private def withEncoder(c: TypedColumn[_, _]): TypedColumn[_, _] = {
```

Comment:
Perhaps this should live in `TypedColumn` itself so we don't have to duplicate the logic.